plugin.video.tvvlaanderen/resources/lib/solocoo/auth.py

```diff
@@ -0,0 +1,388 @@
+# -*- coding: utf-8 -*-
+""" Solocoo Auth API """
+
+from __future__ import absolute_import, division, unicode_literals
+
+import json
+import logging
+import os
+import time
+import uuid
+from hashlib import md5
+
+from requests import HTTPError
+
+from resources.lib.solocoo import TENANTS, SOLOCOO_API, util
+from resources.lib.solocoo.exceptions import InvalidLoginException, InvalidTokenException
+
+try:  # Python 3
+    import jwt
+except ImportError:  # Python 2
+    # The package is named pyjwt in Kodi 18: https://github.com/lottaboost/script.module.pyjwt/pull/1
+    import pyjwt as jwt
+
+try:  # Python 3
+    from urllib.parse import urlparse, parse_qs, urljoin
+except ImportError:  # Python 2
+    from urlparse import urlparse, parse_qs, urljoin
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AccountStorage:
+    """ Data storage for account info """
+    # We will generate a random serial when we don't have any
+    device_serial = ''
+    device_name = 'Kodi'
+
+    # Challenges we can keep to renew our tokens
+    challenge_id = ''
+    challenge_secret = ''
+
+    # Cookie token used to authenticate requests to the app
+    aspx_token = ''
+
+    # Token used to authenticate a request to the tvapi.solocoo.tv endpoint
+    jwt_token = ''
+
+    # Credentials hash
+    hash = ''
+
+    def is_valid_token(self):
+        """ Validate the JWT to see if it's still valid.
+
+        :rtype: boolean
+        """
+        if not self.jwt_token:
+            # We have no token
+            return False
+
+        try:
+            # Verify our token to see if it's still valid.
+            jwt.decode(self.jwt_token,
+                       algorithms=['HS256'],
+                       options={'verify_signature': False, 'verify_aud': False, 'verify_nbf': False})
+        except Exception as exc:  # pylint: disable=broad-except
+            _LOGGER.debug('JWT is NOT valid: %s', exc)
+            return False
+
+        _LOGGER.debug('JWT is valid')
+        return True
+
+
+class AuthApi:
+    """ Solocoo Auth API """
+
+    TOKEN_FILE = 'auth-tokens.json'
+
+    def __init__(self, username, password, tenant, token_path):
+        """ Initialisation of the class.
+
+        :param str username:            The username of the account.
+        :param str password:            The password of the account.
+        :param str tenant:              The tenant code of the account (eg. tvv).
+        :param str token_path:          The path where we can cache our token.
+        """
+        self._username = username
+        self._password = password
+
+        self._tenant = TENANTS.get(tenant)
+        if self._tenant is None:
+            raise Exception('Invalid tenant: %s' % tenant)
+
+        self._token_path = token_path
+
+        # Load existing account data
+        self._account = AccountStorage()
+        self._load_cache()
+
+        # Generate device serial if we have none
+        if not self._account.device_serial:
```

Comment:
Can you explain why you need to expose the user uuid to an external service?

Reply:
To authenticate, we need to present an unique ID for the device to the backend. You can only login to a few devices at the same time, so we need to remember the id that we have generated.

The id we generate is random, and not tied to any system id or something.